File: backend/app/games/minion_clash/js/battle/EntityManager.js
/**
 * Owns all live entities. Provides add/remove/iterate + cull-dead.
 * Entities are referenced by id elsewhere (e.g. projectiles target entities).
 */
export class EntityManager {
    constructor() {
        this._entities = [];
        this._byId = new Map();
        this._toAdd = [];
    }

    add(entity) {
        this._toAdd.push(entity);
    }

    flushAdditions() {
        if (this._toAdd.length === 0) return;
        for (const e of this._toAdd) {
            this._entities.push(e);
            this._byId.set(e.id, e);
        }
        this._toAdd.length = 0;
    }

    getById(id) { return this._byId.get(id); }

    forEach(cb) {
        for (const e of this._entities) cb(e);
    }

    list() { return this._entities; }

    cullDead() {
        let removed = 0;
        for (let i = this._entities.length - 1; i >= 0; i--) {
            const e = this._entities[i];
            if (e.isDead()) {
                this._entities.splice(i, 1);
                this._byId.delete(e.id);
                removed++;
            }
        }
        return removed;
    }
}

File: backend/app/games/minion_clash_be/simulation/entity_manager.py
"""
EntityManager — owns the live entity list with deferred additions and
dead-entity culling. Mirrors `EntityManager.js`.
"""

from __future__ import annotations

from typing import Iterable, Optional

from .entities import Entity


class EntityManager:
    def __init__(self) -> None:
        self._entities: list[Entity] = []
        self._by_id: dict[int, Entity] = {}
        self._to_add: list[Entity] = []

    def add(self, entity: Entity) -> None:
        self._to_add.append(entity)

    def flush_additions(self) -> None:
        if not self._to_add:
            return
        for e in self._to_add:
            self._entities.append(e)
            self._by_id[e.id] = e
        self._to_add.clear()

    def by_id(self, entity_id: int) -> Optional[Entity]:
        return self._by_id.get(entity_id)

    def list(self) -> list[Entity]:
        return self._entities

    def iter_all(self) -> Iterable[Entity]:
        return iter(self._entities)

    def cull_dead(self) -> int:
        if not any(e.is_dead() for e in self._entities):
            return 0
        survivors: list[Entity] = []
        removed = 0
        for e in self._entities:
            if e.is_dead():
                self._by_id.pop(e.id, None)
                removed += 1
            else:
                survivors.append(e)
        self._entities = survivors
        return removed
